### PR TITLE
Adding custom error codes + handling for them

### PR DESCRIPTION
**Pull request by saadberry**
- Added custom error codes `4000` & `4006` for the case: "free plan restrictions"
- Added handling for error in the case when user tries adding more than 1 free repo on FREE plan
- Added new method getUserSubscription that hits `/v1/pricing/subscription` endpoint to GET subscription details
- Added constants for error messages & notification messages

### If (not isOrgRepro and canAvailTrial = False):
image

### If (not isOrgRepro and canAvailTrial = True):
image

### If (isOrgRepro and canAvailTrial = True): 
image

### If (isOrgRepro and canAvailTrial = False):
image

### CODE REVIEW

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java [hunk 1]
```diff
@@ -95,6 +98,37 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+//                        try {
+//                            return validateAccessToken(accessToken);
+//                        } catch (InvalidAccessTokenError invalidAccessTokenError) {
+//                            // no action needed, user will be prompted for authenticated later by default.
+//                        }
+//                    }
+                    // Hit /v1/init endpoint
+                    JSONResponse jsonResponse;
+                    try {
+                        jsonResponse = ClientUtils.sendPost(API_INIT, accessToken);
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java:120]: @saadberry Aren't we doing this already in the code somewhere? The code that "Connects a Repo". We need to add error handling on the existing API call

File: src/main/java/org/intellij/sdk/codesync/Constants.java [hunk 2]
```diff
@@ -226,4 +226,18 @@ private ErrorCodes() {
...
+    public static final class CustomErrorCodes {
+        /*
+            Custom error codes for the case: Free Plan restrictions
+        */
+        private CustomErrorCodes(){
+            // restrict instantiation
+        }
+
+        public static final int IS_FROZEN_REPO = 4000;
+        public static final int REPO_SIZE_LIMIT_REACHED = 4001;
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/Constants.java:239]: I think we this PR targets ```PRIVATE_REPO_COUNT_LIMIT_REACHED = 4006``` error code instead of these

File: src/main/java/org/intellij/sdk/codesync/Constants.java [hunk 3]
```diff
@@ -180,6 +180,9 @@ private Notification() {
         public static final String TRY_PRO_FOR_FREE = "To continue, please sign up for a 30-day free trial of Pro.";
         public static final String TRY_ORG_PRO_FOR_FREE = "To continue, please sign your organization up for a 30-day free trial of the Team plan.";
 
+        // Free plan related notification messages
+        public static final String PRIVATE_REPO_COUNT_REACHED = "Only 1 private repository is allowed in Free Plan";
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/Constants.java:184]: How about? (GPT suggestion)
```"In the Free Plan, you can have just one private repository."```

File: src/main/java/org/intellij/sdk/codesync/Constants.java (hunk 2 above)
Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/Constants.java:252]: I think above comment still needs to be addressed @saadberry

File: src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java [hunk 4]
```diff
@@ -27,10 +28,27 @@ public PricingAlertDialog(Boolean isOrgRepo, Boolean canAvailTrial, String prici
         this.pricingURL = pricingURL;
 
         if (canAvailTrial) {
-            upgradeButtonText = "Try Pro for free";
+            upgradeButtonText = isOrgRepo ? NotificationButton.TRY_TEAM_FOR_FREE: NotificationButton.TRY_PRO_FOR_FREE;
             secondaryMessage = isOrgRepo ? Notification.TRY_ORG_PRO_FOR_FREE: Notification.TRY_PRO_FOR_FREE;
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java:32]: Not part of this PR but we can rename TRY_ORG_PRO_FOR_FREE to TRY_TEAM_PLAN_FOR_FREE

File: src/main/java/org/intellij/sdk/codesync/exceptions/response/StatusCodeError.java [hunk 5]
```diff
@@ -6,12 +6,24 @@
...
         this.statusCode = statusCode;
+        this.customErrorCode = customErrorCode;
     }
 
     public int getStatusCode() {
         return statusCode;
     }
+
+    public int getCustomErrorCode() {
+        System.out.println("in custom error code implementation");
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/exceptions/response/StatusCodeError.java:26]: Need to remove all print statements

File: src/main/resources/META-INF/plugin.xml [hunk 6]
```diff
@@ -94,19 +94,9 @@
...
-        <li>Handling for disabled accounts and the ability to enable it again.</li>
-        <li>Integration with Github pull request to watch pull request playback.</li>
-    <h2>Bug Fixes</h2>
-      <ul>
-        <li>Fixed an error in handle buffer. Added checks for null values.</li>
-        <li>Updated messaging for repo disconnect confirmation prompt.</li>
-        <li>Only process diffs of the authenticated user.</li>
-        <li>Disabled config file reset logic in case of yaml error.</li>
+        <li>Added new error message "Only 1 private repository is allowed in Free Plan"</li>
+        <li>Added custom error codes class: CustomErrorCodes</li>
```
Comment (Basit-qc) [inline src/main/resources/META-INF/plugin.xml:98]: This is shown to the end user, so we should be descriptive about it, should not talk much about technical details

Comment (Basit-qc) [reply src/main/resources/META-INF/plugin.xml:98]: Also, we can keep previous lines for next couple releases since it highlights recent CHANGELOG

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java [hunk 7]
```diff
@@ -95,6 +98,37 @@ public void actionPerformed(@NotNull AnActionEvent e) {
                     // Show Pricing alert dialog.
                     PricingAlerts.setPlanLimitReached(project);
                 } else {
+                    // Get access token
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java:101]: Why do we have this block of code? We don't need any separate API call except the existing REPO-INIT call, right?

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java [hunk 8]
```diff
@@ -182,6 +183,8 @@ public static void setupCodeSyncRepoAsync(Project project, String repoPath, Stri
         StateUtils.updateRepoStatus(repoPath, RepoStatus.SYNC_IN_PROGRESS);
         ProgressManager.getInstance().run(new Task.Backgroundable(project, "Initializing repo"){
             public void run(@NotNull ProgressIndicator progressIndicator) {
+
+
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:187]: Kindly remove redundant lines

**Review by Basit-qc — CHANGES_REQUESTED** [review]
Looks great @saadberry , requested some changes, kindly have a look, thanks

File: src/main/java/org/intellij/sdk/codesync/Constants.java (hunk 2 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/Constants.java:252]: I had kept them for future cases but i have realized there will be no case to send 4001 from plugin. I shall remove it then. But there can be a case for sending 4000, in case a user tries to modify a frozen repo right? So we should keep that

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java [hunk 9]
```diff
@@ -34,6 +34,20 @@ public static void setPlanLimitReached(Project project) {
         );
     }
 
+    public static void setPlanLimitReached(Project project, boolean privateRepoCountLimitReached) {
+        // We only want to show notification once every 5 minutes, I have implemented that using locks with an expiry of
+        // 5 minutes. So, skip the notification if lock is not acquired.
+        acquirePricingLock();
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:40]: I don't think we need locking here since we want to show notiification only once

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java [hunk 10]
```diff
@@ -34,6 +34,20 @@ public static void setPlanLimitReached(Project project) {
...
+        CommonUtils.invokeAndWait(
+                () -> {
+                    PricingAlertDialog pricingAlertDialog = new PricingAlertDialog(false, false, CODESYNC_PRICING_URL, project, privateRepoCountLimitReached);
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:43]: Why are we passing hardcoded false for `canAvailTrial`? Shouldn't this come from the API? `/v1/users/pricing/subscription` (I think)

File: src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java [hunk 11]
```diff
@@ -27,10 +28,27 @@ public PricingAlertDialog(Boolean isOrgRepo, Boolean canAvailTrial, String prici
...
         } else {
-            upgradeButtonText = "Upgrade to Pro";
+            upgradeButtonText = NotificationButton.UPGRADE_PLAN;
+            secondaryMessage = isOrgRepo ? Notification.UPGRADE_ORG_PRICING_PLAN: Notification.UPGRADE_PRICING_PLAN;
+        }
+
+        setTitle(this.title);
+    }
+
+    public PricingAlertDialog(Boolean isOrgRepo, Boolean canAvailTrial, String pricingURL, Project project, boolean privateRepoCountLimitReached){
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java:41]: We probably can keep only 1 param here `boolean privateRepoCountLimitReached` and then can above overloaded method with 
1. isOrgRepo = false
2. canAvailTrial will come from the API
3. pricingURL will come from the constants

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java (hunk 9 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:40]: showPrivateRepoCountReached

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java [hunk 12]
```diff
@@ -34,6 +34,20 @@ public static void setPlanLimitReached(Project project) {
         );
     }
 
+    public static void setPlanLimitReached(Project project, boolean privateRepoCountLimitReached) {
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:37]: We can rename this to `showPrivateRepoCountLimitReached`, we don't need any param here. 
We will do 
1. `getUserSubscription()` and call ```public PricingAlertDialog(Boolean canAvailTrial)``` with API response

File: src/main/java/org/intellij/sdk/codesync/clients/JSONResponse.java [hunk 13]
```diff
@@ -17,10 +17,17 @@
 */
 public class JSONResponse {
     private final int statusCode;
+    private int customErrorCode;
```
Comment (saleem-latif) [inline src/main/java/org/intellij/sdk/codesync/clients/JSONResponse.java:20]: I do not this one being used anywhere. I think we should not make this a property.

**Review by saleem-latif — CHANGES_REQUESTED** [review]

File: src/main/java/org/intellij/sdk/codesync/clients/JSONResponse.java (hunk 13 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/clients/JSONResponse.java:20]: We need this to store value of custom error code 
image

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java [hunk 14]
```diff
@@ -34,6 +34,21 @@ public static void setPlanLimitReached(Project project) {
         );
     }
 
+    public static void showPrivateRepoCountLimitReached() {
+            // Validate that user can avail trial
+            Boolean  canAvailTrial = CodeSyncClient.getUserSubscription();
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:39]: nit; Seems like 2 spaces after the Type declaration

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java [hunk 15]
```diff
@@ -34,6 +34,21 @@ public static void setPlanLimitReached(Project project) {
...
+            System.out.println(canAvailTrial);
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:40]: Kindly remove print statements, thanks

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java [hunk 16]
```diff
@@ -271,4 +277,20 @@ public JSONObject getTeamActivity(String accessToken) {
         }
     }
 
+    public  static  Boolean  getUserSubscription() {
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:280]: Why do we have double spaces between these words?

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java [hunk 17]
```diff
@@ -271,4 +277,20 @@ public JSONObject getTeamActivity(String accessToken) {
...
+        */
+        String accessToken = UserAccount.getAccessTokenByEmail();
+        try {
+            JSONResponse response = ClientUtils.sendGet(USER_SUBSCRIPTION_ENDPOINT, accessToken);
+            JSONObject result = response.getJsonResponse();
+            JSONObject subscriptionObject = (JSONObject) result.get("subscription");
+            return Boolean.valueOf(subscriptionObject.get("can_avail_trial").toString());
+        } catch (RequestError | InvalidJsonError | StatusCodeError error) {
+            CodeSyncLogger.error(String.format("Error while getting user subscription. %s", error.getMessage()));
+            return null;
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:292]: Since ouput type is boolean, maybe we should return false here?

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java [hunk 18]
```diff
@@ -13,6 +13,7 @@
 import org.intellij.sdk.codesync.CodeSyncLogger;
 import org.intellij.sdk.codesync.NotificationManager;
 import org.intellij.sdk.codesync.auth.CodeSyncAuthServer;
+import org.intellij.sdk.codesync.clients.ClientUtils;
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:16]: What is this import doing?

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java [hunk 19]
```diff
@@ -8,13 +8,15 @@
 import org.intellij.sdk.codesync.NotificationManager;
 import org.intellij.sdk.codesync.actions.BaseModuleAction;
 import org.intellij.sdk.codesync.alerts.PricingAlerts;
-import org.intellij.sdk.codesync.exceptions.InvalidConfigFileError;
-import org.intellij.sdk.codesync.exceptions.SQLiteDBConnectionError;
-import org.intellij.sdk.codesync.exceptions.SQLiteDataError;
+import org.intellij.sdk.codesync.clients.ClientUtils;
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java:11]: Why do we have only imports changed?

File: src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java [hunk 20]
```diff
@@ -25,18 +29,22 @@ public class PricingAlertDialog extends DialogWrapper {
...
-            upgradeButtonText = "Upgrade to Pro";
+            upgradeButtonText = NotificationButton.UPGRADE_PLAN;
             secondaryMessage = isOrgRepo ? Notification.UPGRADE_ORG_PRICING_PLAN: Notification.UPGRADE_PRICING_PLAN;
         }
 
         setTitle(this.title);
     }
 
+    public PricingAlertDialog(Boolean canAvailTrial){
+        this(true, false, CODESYNC_PRICING_URL, null);
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java:44]: Where are we using canAvailTrial? 🤔 Seems like we are passing hard coded False
Also, why are we passing true for isOrgRepo?

**Review by Basit-qc — CHANGES_REQUESTED** [review]

File: src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java (hunk 20 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java:44]: Oh yes, I was passing different values for our cases. Forgot to revert. Updating

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java (hunk 18 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:16]: Believe we had added this import in our previous implementations, we are not using it now. Removing

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java (hunk 16 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:280]: Not quite sure how that came. Removing

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java (hunk 17 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:292]: Good idea

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java (hunk 19 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java:11]: Believe we were implementing our old logic here. The implementations were removed but the imports were not. Thanks for pointing this out. Removing

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java [hunk 21]
```diff
@@ -222,9 +223,14 @@ public JSONObject uploadRepo(String accessToken, JSONObject payload) {
             CodeSyncLogger.critical(String.format("Error while repo init, %s", error.getMessage()));
             return null;
         } catch (StatusCodeError statusCodeError) {
-            if (statusCodeError.getStatusCode()  == ErrorCodes.REPO_SIZE_LIMIT_REACHED) {
+
+            if (statusCodeError.getCustomErrorCode()  == CustomErrorCodes.PRIVATE_REPO_COUNT_LIMIT_REACHED) {
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:227]: nit, remove extra space before `==` @saadberry. 
You can run Format Code in the Intellij IDEA to fix such minor syntax issues

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java [hunk 22]
```diff
@@ -222,9 +223,14 @@ public JSONObject uploadRepo(String accessToken, JSONObject payload) {
...
+                PricingAlerts.showPrivateRepoCountLimitReached();
+            }
+            else if (statusCodeError.getStatusCode()  == ErrorCodes.REPO_SIZE_LIMIT_REACHED) {
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:230]: This is generally catch all cases where status code = 402 so we should use this carefully

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java (hunk 21 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:227]: Thanks. I will use that to check quality in the future

File: src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java (hunk 20 above)
Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/ui/dialogs/PricingAlertDialog.java]: I see, we should be testing with actual code and create data on the server we need for local testing so that actual code is tested on different cases

File: src/main/resources/META-INF/plugin.xml [hunk 23]
```diff
@@ -94,6 +94,8 @@
   <change-notes><![CDATA[
     <h2>New Features</h2>
       <ul>
+        <li>Restricted users on free plan to only one private repository</li>
```
Comment (Basit-qc) [inline src/main/resources/META-INF/plugin.xml:97]: I think @saleem-latif  was adding to the bottom of this list so probably should use same (I am fine either way but it should be same for every update)

**Review by Basit-qc — APPROVED** [review]

File: src/main/resources/META-INF/plugin.xml (hunk 23 above)
Comment (saadberry) [reply src/main/resources/META-INF/plugin.xml:97]: Yeah let's follow the convention that has been being used originally

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java [hunk 24]
```diff
@@ -18,25 +18,41 @@ private static void acquirePricingLock() {
...
         // 5 minutes. So, skip the notification if lock is not acquired.
         acquirePricingLock();
         CommonUtils.invokeAndWait(
-            () -> {
-                PricingAlertDialog pricingAlertDialog = new PricingAlertDialog(false, false, CODESYNC_PRICING_URL, project);
-                return pricingAlertDialog.showAndGet();
-            },
-            ModalityState.defaultModalityState()
+                () -> {
+                    PricingAlertDialog pricingAlertDialog = new PricingAlertDialog(false, false, CODESYNC_PRICING_URL, project);
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:32]: Do we know why are we passing hardcoded false for canAvailTrial here @saadberry ?

File: src/main/resources/META-INF/plugin.xml [hunk 25]
```diff
@@ -101,6 +101,8 @@
         <li>Async file upload to speed up repo setup.</li>
         <li>Handling for disabled accounts and the ability to enable it again.</li>
         <li>Integration with Github pull request to watch pull request playback.</li>
+         <li>Restricted users on free plan to only one private repository</li>
```
Comment (Basit-qc) [inline src/main/resources/META-INF/plugin.xml:104]: nit, kindly remove redundant space before starting tag @saadberry

File: src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java (hunk 24 above)
Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:32]: I believe that is the default case when we cannot assess wether user is eligible for trial or not

Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:32]: Alright, can you kindly open a ticket for this to use actual API call here since you have added that API call now

Comment (saadberry) [reply src/main/java/org/intellij/sdk/codesync/alerts/PricingAlerts.java:32]: Done: https://linear.app/codesync/issue/PLU-8/fetch-canavailtrial-from-subscription-api

File: src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java [hunk 26]
```diff
@@ -41,7 +43,7 @@ public Boolean isServerUp() {
             return false;
         } catch (StatusCodeError error) {
             CodeSyncLogger.error(
-                String.format("Server returned %s status code on health check endpoint.", error.getStatusCode())
+                    String.format("Server returned %s status code on health check endpoint.", error.getStatusCode())
```
Comment (saleem-latif) [inline src/main/java/org/intellij/sdk/codesync/clients/CodeSyncClient.java:46]: I have been using 3 space indent for almost all files in the code. Can you switch to that as well?

File: src/main/java/org/intellij/sdk/codesync/clients/JSONResponse.java (hunk 13 above)
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/clients/JSONResponse.java:20]: I think that can be converted to a local variable.

**Review by saleem-latif — APPROVED** [review]
A couple of comments, other than that looks good 👍🏼
